cardano-cli/src/Cardano/CLI/EraBased/Run/Governance/GenesisKeyDelegationCertificate.hs

```diff
@@ -0,0 +1,69 @@
+{-# LANGUAGE GADTs #-}
+
+module Cardano.CLI.EraBased.Run.Governance.GenesisKeyDelegationCertificate (runGovernanceGenesisKeyDelegationCertificate) where
+
+
+import           Cardano.Api
+import           Cardano.Api.Shelley
+
+import           Cardano.CLI.Types.Errors.GovernanceCmdError
+import           Cardano.CLI.Types.Key
+
+import           Control.Monad.Trans.Except (ExceptT)
+import           Control.Monad.Trans.Except.Extra
+
+runGovernanceGenesisKeyDelegationCertificate
+  :: ShelleyBasedEra era
+  -> VerificationKeyOrHashOrFile GenesisKey
+  -> VerificationKeyOrHashOrFile GenesisDelegateKey
+  -> VerificationKeyOrHashOrFile VrfKey
+  -> File () Out
+  -> ExceptT GovernanceCmdError IO ()
+runGovernanceGenesisKeyDelegationCertificate sbe
+                                             genVkOrHashOrFp
+                                             genDelVkOrHashOrFp
+                                             vrfVkOrHashOrFp
+                                             oFp = do
+  genesisVkHash <- firstExceptT GovernanceCmdKeyReadError
+    . newExceptT
+    $ readVerificationKeyOrHashOrTextEnvFile AsGenesisKey genVkOrHashOrFp
+  genesisDelVkHash <-firstExceptT GovernanceCmdKeyReadError
+    . newExceptT
+    $ readVerificationKeyOrHashOrTextEnvFile AsGenesisDelegateKey genDelVkOrHashOrFp
+  vrfVkHash <- firstExceptT GovernanceCmdKeyReadError
+    . newExceptT
+    $ readVerificationKeyOrHashOrFile AsVrfKey vrfVkOrHashOrFp
+
+  req <- hoistEither $ createGenesisDelegationRequirements sbe genesisVkHash genesisDelVkHash vrfVkHash
+  let genKeyDelegCert = makeGenesisKeyDelegationCertificate req
+
+  firstExceptT GovernanceCmdTextEnvWriteError
+    . newExceptT
+    $ writeLazyByteStringFile oFp
+    $ shelleyBasedEraConstraints sbe
+    $ textEnvelopeToJSON (Just genKeyDelegCertDesc) genKeyDelegCert
+  where
+    genKeyDelegCertDesc :: TextEnvelopeDescr
+    genKeyDelegCertDesc = "Genesis Key Delegation Certificate"
+
+createGenesisDelegationRequirements
+  :: ShelleyBasedEra era
+  -> Hash GenesisKey
+  -> Hash GenesisDelegateKey
+  -> Hash VrfKey
+  -> Either GovernanceCmdError (GenesisKeyDelegationRequirements era)
+createGenesisDelegationRequirements sbe hGen hGenDeleg hVrf =
+  case sbe of
+    ShelleyBasedEraShelley -> do
+      return $ GenesisKeyDelegationRequirements ShelleyToBabbageEraShelley hGen hGenDeleg hVrf
+    ShelleyBasedEraAllegra -> do
+      return $ GenesisKeyDelegationRequirements ShelleyToBabbageEraAllegra hGen hGenDeleg hVrf
+    ShelleyBasedEraMary -> do
+      return $ GenesisKeyDelegationRequirements ShelleyToBabbageEraMary hGen hGenDeleg hVrf
+    ShelleyBasedEraAlonzo -> do
+      return $ GenesisKeyDelegationRequirements ShelleyToBabbageEraAlonzo hGen hGenDeleg hVrf
+    ShelleyBasedEraBabbage -> do
+      return $ GenesisKeyDelegationRequirements ShelleyToBabbageEraBabbage hGen hGenDeleg hVrf
+    ShelleyBasedEraConway ->
+      Left GovernanceCmdGenesisDelegationNotSupportedInConway
```

Comment:
You should be able to use `caseShelleyToBabbageOrConwayEraOnwards` instead of `case` here.

Reply:
Thanks @newhoggy, done :+1: